apicast/src/configuration/service.lua

```diff
@@ -0,0 +1,167 @@
+local setmetatable = setmetatable
+local tostring = tostring
+local rawget = rawget
+local next = next
+local lower = string.lower
+local gsub = string.gsub
+local select = select
+local type = type
+
+local http_authorization = require 'resty.http_authorization'
+
+local _M = {
+
+}
+
+local mt = {
+  __index = _M
+}
+
+function _M.new(attributes)
+  return setmetatable(attributes or {}, mt)
+end
+
+local http_methods_with_body = {
+  POST = true,
+  PUT = true,
+  PATCH = true
+}
+
+local function read_body_args(...)
+  local method = ngx.req.get_method()
+
+  if not http_methods_with_body[method] then
+    return {}, 'not supported'
+  end
+
+  ngx.req.read_body()
+
+  local args = ngx.req.get_post_args()
+  local results = {}
+
+  for n=1, select('#', ...) do
+    results[n] = args[select(n, ...)]
+  end
+
+  return results
+end
+
+local function read_http_header(name)
+  local normalized = gsub(lower(name), '-', '_')
+  return ngx.var['http_' .. normalized]
+end
+
+local credentials_mt = {
+  -- nipairs = only non integer pairs
+  __pairs = function (t)
+    return function(_, k)
+      local v
+      repeat
+        k, v = next(t, k)
+      until k == nil or type(k) ~= 'number'
+      return k, v
+    end, t, nil
+  end
+}
+
+local backend_version_credentials = {
```

Comment:
I think that extracting one method for each of the versions would help readability. What do you think @mikz ?